### PR TITLE
Infinispan 15: indexing and query news

### PR DESCRIPTION
**Pull request by fax4ever**

### CODE REVIEW

File: _posts/2024-06-10-infinispan-15-queries.adoc [hunk 1]
```diff
@@ -0,0 +1,317 @@
...
+
+[[rest_projections]]
+== Rest queries with projections
+
+Recently we introduced score and versions projections, that along with entity and field projections are the all
+kinds of projections we support with the Infinispan queries.
+
+The same projections are now available using the REST query API.
+
+For instance if we have an entity like the following:
```
Comment (rigazilla) [inline _posts/2024-06-10-infinispan-15-queries.adoc:27]: I would group entity, query and result... something like `Following an example of entity, query and result:`

File: _posts/2024-06-10-infinispan-15-queries.adoc [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+        }
+    ]
+}
+----
+
+
+In this case we have requested the version projection using the projection function `version(g)`
+and receiving the corresponding result in the `version()` attribute of the returned `hit`.
+Similarly, a `score()` projection is produced.
+The special value `*` corresponds to the entity projection result, in which the special field `_type` corresponds to the type of the entity.
```
Comment (rigazilla) [inline _posts/2024-06-10-infinispan-15-queries.adoc:96]:
```suggestion
The special attribute `*` corresponds to the entity projection result, in which the special field `_type` corresponds to the type of the entity.
```

File: _posts/2024-06-10-infinispan-15-queries.adoc [hunk 3]
```diff
@@ -0,0 +1,317 @@
...
+
+
+[[continuous-queries]]
+== Query cache API
+
+The use of the search factory to access the query APIs is no longer required and it is now deprecated.
+
+From Infinispan 15 we can simply create a query instance from both a remote or an embedded cache (the API is the same)
+using the method `cache#query`.
```
Comment (rigazilla) [inline _posts/2024-06-10-infinispan-15-queries.adoc:105]:
```suggestion
With Infinispan 15, both embedded and remote caches can be queried with the same API metho `cache#query`. The use of the search factory to access the query APIs is no longer required and it is now deprecated.
```

File: _posts/2024-06-10-infinispan-15-queries.adoc [hunk 4]
```diff
@@ -0,0 +1,317 @@
...
+
+[source,java]
+----
+query.startOffset(10);
+query.maxResults(10);
+QueryResult<Person> = query.execute();
+----
+
+Similarly, it is possible to get a continuous query instance from the same cache (both remote and embedded),
+using the method `cache#continuousQuery`:
```
Comment (rigazilla) [inline _posts/2024-06-10-infinispan-15-queries.adoc:123]:
```suggestion
Similarly, it is possible to get a continuous query instance (both for remote and embedded cache), using the method `cache#continuousQuery`:
```

File: _posts/2024-06-10-infinispan-15-queries.adoc [hunk 5]
```diff
@@ -0,0 +1,317 @@
...
+
+[source,java]
+----
+ContinuousQuery<Integer, Person> continuousQuery = myCache.continuousQuery();
+----
+
+From which as usual it will be possible to define the continuous callback:
```
Comment (rigazilla) [inline _posts/2024-06-10-infinispan-15-queries.adoc:130]: I would remove this and group the example

File: _posts/2024-06-10-infinispan-15-queries.adoc [hunk 6]
```diff
@@ -0,0 +1,317 @@
...
+      public void resultLeaving(Integer key) {
+         // handle entry leavings
+      }
+   });
+----
+
+[[filter-elements]]
+== Filter elements for kNN queries
+
+kNN queries can be run filtering the population on which to apply the search.
```
Comment (rigazilla) [inline _posts/2024-06-10-infinispan-15-queries.adoc:154]:
```suggestion
kNN queries can run on a pre-filtered subset of the population.
```

File: _posts/2024-06-10-infinispan-15-queries.adoc [hunk 7]
```diff
@@ -0,0 +1,317 @@
...
+A kNN filter is defined using any kind of predicate (included boolean expressions) provided by the Infinispan queries.
```
Comment (rigazilla) [inline _posts/2024-06-10-infinispan-15-queries.adoc:155]:
```suggestion
A kNN filter is defined using any kind of predicate (included boolean expressions) provided by the Infinispan query language.
```

File: _posts/2024-06-10-infinispan-15-queries.adoc [hunk 8]
```diff
@@ -0,0 +1,317 @@
...
+   @ProtoField(3)
+   public String getDescription() {
+      return description;
+   }
+}
+----
+
+which is mapped using different indexing annotations (`@Keyword`, `@Text`, `@Vector`, ...).
+
+We can limit the population on which to apply the search to the only items having a description text containing the term `cat`.
```
Comment (rigazilla) [inline _posts/2024-06-10-infinispan-15-queries.adoc:197]:
```suggestion
Suppose that we want to limit the vector search only to record with the word 'cat' in the description. We can do like this:
```

File: _posts/2024-06-10-infinispan-15-queries.adoc [hunk 9]
```diff
@@ -0,0 +1,317 @@
...
+      "select score(i), i from Item i where i.floatVector <-> [:a]~:k filtering i.description : 'cat'");
+query.setParameter("a", new float[]{7.0f, 7.0f, 7.0f});
+query.setParameter("k", 3);
+
+List<Object[]> hits = query.list();
+----
+
+In particular, this is also an example of how to combine full text search and vector search.
+
+Also boolean composite predicates can be used to filter the search.
```
Comment (rigazilla) [inline _posts/2024-06-10-infinispan-15-queries.adoc:211]:
```suggestion
This is example shows a combination of full text search and vector search, but boolean composite predicates are also supported.
```

File: _posts/2024-06-10-infinispan-15-queries.adoc [hunk 10]
```diff
@@ -0,0 +1,317 @@
...
+In the following we will limit the search to the only items having the term `cat` in their description and having code `w739`.
```
Comment (rigazilla) [inline _posts/2024-06-10-infinispan-15-queries.adoc:212]:
```suggestion
In the following we will limit the search only to the items having the term `cat` in their description and having code `w739`.
```

**Review by rigazilla — COMMENTED** [review]
The first commit is intentional?
added some suggestions, don't trust my english anyway :)

File: _posts/2024-06-10-infinispan-15-queries.adoc [hunk 11]
```diff
@@ -0,0 +1,317 @@
...
+title: Infinispan 15 indexing & query news
+permalink: /blog/:year/:month/:day/infinispan-15-queries
+date: '2024-06-10T00:00:00.000-00:00'
+author: fax4ever
+tags: [ "search", "vector", "score", "knn", "indexing", "embeddings" ]
+---
+
+= Infinispan 15 indexing & query news
+
+The Infinispan team is happy to announce some improvements on query APIs that are already present on Infinispan 15:
```
[the remark this answers is not in the document]
Comment (tristantarrant) [reply _posts/2024-06-10-infinispan-15-queries.adoc:12]: @rigazilla is right. This should be more like "A short while back we released Infinispan 15 which delivered many improvements to the query API. This blog is an in-depth dive into some of these"

[the remark this answers is not in the document]
Comment (fax4ever) [reply _posts/2024-06-10-infinispan-15-queries.adoc:12]: sure! I'm applying the changes

File: _posts/2024-06-10-infinispan-15-queries.adoc (hunk 5 above)
Comment (fax4ever) [reply _posts/2024-06-10-infinispan-15-queries.adoc:130]: I would prefer to keep it, to see where the story ends, even if we didn't changed this API.

File: _posts/2024-06-10-infinispan-15-queries.adoc (hunk 6 above)
Comment (fax4ever) [reply _posts/2024-06-10-infinispan-15-queries.adoc:154]: I prefer the original from, I think it is a little more expressive.

**Comment by fax4ever** [comment]
Thanks a lot @rigazilla for the review. I added you as coauthor of the commit.
Yes the first commit is intentional and without it I have some issue on running the jekyll server with Fedora 40

**Review by rigazilla — APPROVED** [review]
lgtm!

**Pull request merged**